src/System.IO.Pipelines/System/IO/Pipelines/ReadableBuffer.cs

```diff
@@ -9,24 +9,23 @@ namespace System.IO.Pipelines
     /// <summary>
     /// Represents a buffer that can read a sequential series of bytes.
     /// </summary>
-    public struct ReadableBuffer
+    public readonly struct ReadableBuffer
     {
-        internal ReadCursor BufferStart;
-        internal ReadCursor BufferEnd;
-        internal long BufferLength;
+        internal readonly ReadCursor BufferStart;
+        internal readonly ReadCursor BufferEnd;
 
         /// <summary>
         /// Length of the <see cref="ReadableBuffer"/> in bytes.
         /// </summary>
-        public long Length => BufferLength;
+        public long Length => BufferStart.GetLength(BufferEnd);
 
         /// <summary>
         /// Determines if the <see cref="ReadableBuffer"/> is empty.
         /// </summary>
-        public bool IsEmpty => BufferLength == 0;
+        public bool IsEmpty => Length == 0;
```

Comment:
Does removing actual length cause issue for `IsEmpty` as its now a calculation? 

Can it just check the first segment rather than full chain?

Reply:
We need to figure out some issues with cursors and switch to cursor comparison for `IsEmpty`